File: snippets/cpp/minimal_trimap.cpp
/**
 * @brief  Minimal GrabCut segmentation code snippet.
 *
 * @author Luis Carlos Garcia Peraza Herrera (luiscarlos.gph@gmail.com).
 * @date   11 Feb 2021.
 */

#include <opencv2/core/core.hpp>
#include <grabcut/grabcut.h>
//#include <opencv2/highgui/highgui.hpp>
//#include <opencv2/imgproc/imgproc.hpp>

int main(int argc, char **argv) {
  const std::string imagePath = "data/tool_512x409.png"; 
  const std::string trimapPath = "data/trimap_512x409.png";
  const std::string outputPath = "data/output_512x409_trimap_iter_5_gamma_10.png";
  int maxIter = 5;
  float gamma = 10.;

  // Read image and trimap
  cv::Mat im = cv::imread(imagePath);
  cv::Mat imBgra;
  cv::cvtColor(im, imBgra, cv::COLOR_BGR2BGRA);
  cv::Mat trimap = cv::imread(trimapPath, cv::IMREAD_GRAYSCALE);

  // Perform segmentation
  GrabCut gc(maxIter);
  cv::Mat segmentation = gc.estimateSegmentationFromTrimap(imBgra, trimap, gamma);

  // Save segmentation
  cv::imwrite(outputPath, segmentation);

  return 0;
}

File: snippets/python/minimal_trimap.py
#!/usr/lib/python
#
# @brief Minimal code snippet to show GrabCut segmentation with a trimap.
#
# @author Luis C. Garcia Peraza Herrera (luiscarlos.gph@gmail.com).
# @date   11 Feb 2021.

import cv2
import grabcut

image_path = 'data/tool_512x409.png' 
trimap_path = 'data/trimap_512x409.png'
output_path = 'data/output_512x409_trimap_iter_5_gamma_10.png'
max_iter = 5
gamma = 10.

# Read image and trimap
im = cv2.imread(image_path)
im_bgra = cv2.cvtColor(im, cv2.COLOR_BGR2BGRA)
trimap = cv2.imread(trimap_path, cv2.IMREAD_GRAYSCALE)

# Perform segmentation
gc = grabcut.GrabCut(max_iter)
segmentation = gc.estimateSegmentationFromTrimap(im_bgra, trimap, gamma)

# Save segmentation
cv2.imwrite(output_path, segmentation)
